wiring/inc/spark_wiring_system.h

```diff
@@ -105,6 +105,164 @@ inline SleepResult::SleepResult(WakeupReason r, system_error_t e, pin_t p)
       pin_(p) {
 }
 
+/**
+ * 1. CPU is stopped
+ * 2. Network is stopped
+ * 3. Internal/external peripherals are stopped
+ * 4. RAM memory is not retained
+ */
+enum class SystemSleepMode: uint16_t {
+    NONE        = SYSTEM_SLEEP_MODE_NONE,
+    STOP        = SYSTEM_SLEEP_MODE_STOP,
+    HIBERNATE   = SYSTEM_SLEEP_MODE_HIBERNATE,
+    SHUTDOWN    = SYSTEM_SLEEP_MODE_SHUTDOWN
+};
+
+class SystemSleepConfiguration {
+public:
+    SystemSleepConfiguration() {
+        config_.size = sizeof(hal_core_sleep_config_t);
+        config_.version = HAL_CORE_SLEEP_VERSION;
+        config_.mode = SYSTEM_SLEEP_MODE_NONE;
+        config_.wakeup_sources = nullptr;
+    }
+    ~SystemSleepConfiguration() {
+        // Free memory
+        auto wakeupSource = config_.wakeup_sources;
+        while (!wakeupSource) {
+            auto next = wakeupSource->next;
+            delete wakeupSource;
+            wakeupSource = next;
+        }
+    }
+
+    // Getters
+    hal_core_sleep_config_t* halCoreSleepConfig() const {
+        return &config_;
+    }
+
+    SystemSleepMode sleepMode() const {
+        return static_cast<SystemSleepMode>(config_.mode);
+    }
+
+    bool wakeupSourceFeatured(hal_core_wakeup_source_type_t type) const {
```

Comment:
I would also suggest adding `start` argument, so that we can continue the search from a certain position. E.g. if we want to check for duplicate gpio entries and change the return type to a pointer to an actual wakeup source instead of just `bool`, so we can inspect its contents.